### PR TITLE
Native QuadTreeLod

### PR DESCRIPTION
**Pull request by tinmanjuggernaut**
This PR ports QuadTreeLod to C++ GDNative with a GDScript fallback. 

@painfulexistence helped me fix some bugs in the implementation.

On a 2k terrain with a GTX 1060, windows 10/64, I get about a 5-20% improvement in FPS depending on where I look.

### CODE REVIEW

**Comment by Zylann** [comment]
Note, I completely stopped doing things with GDNative as I am waiting for GDExtensions in Godot 4 now, which may be a better system.

I didnt have a look at the code yet, but first obvious question would be, did you compile with optimizations enabled? (i.e no debug)

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+            Godot::print("Error: QuadtreeLod::_update: _vertical_bounds_func did not return a vector2");
+        world_center.y = (vbounds.x + vbounds.y) / 2.0;
+    }
+
+    int split_distance = _base_size * lod_factor * _split_scale;
+
+    if (! quad->has_children()) {
+        if (lod > 0 && world_center.distance_to(view_pos) < split_distance) {
+            // Split
+            quad->add_children();
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:120]: add_children() creates new instances of `Quad` inside the quad. But in the loop below, you are replacing these quads with yet another set of Quad instances, leaking the previous ones

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+} // _update
+
+void QuadTreeLod::_join_all_recursively(Quad *quad, int lod) {
+    if (quad->has_children()) {
+        for (int i = 0; i < 4; i++) {
+            _join_all_recursively(quad->get_child(i), lod - 1);
+        }
+        quad->clear_children();
+
+    } else if (quad->data.get_type() == Variant::OBJECT) {
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:172]: The original check was != null

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+
+        int chunk_indicator = 0;
+        if (quad->data.get_type() == Variant::OBJECT)
+            chunk_indicator = 1;
+
+        Variant rect2(Rect2( Vector2((real_t)quad->origin_x, (real_t)quad->origin_y) * (real_t)size, 
+                             Vector2((real_t)size, (real_t)size)));
+        Variant color(Color(1.0 - (real_t)lod_index * 0.2, 0.2 * (real_t)checker, (real_t)chunk_indicator, 1));
+        Variant *args[] = { &rect2, &color };
+        ci.call("draw_rect", (const Variant **)args, 2);
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:215]: Instead of using `call` which allocates a string everytime and looks up the method dynamically, you may expect a `CanvasItem*` instead of `Variant` so you can call it directly

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+        }
+
+    } else {
+        int size = get_lod_factor(lod_index);
+        int checker = 0;
+        if (child_index == 1 || child_index == 2)
+            checker = 1;
+
+        int chunk_indicator = 0;
+        if (quad->data.get_type() == Variant::OBJECT)
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:208]: The original check was != null

File: addons/zylann.hterrain/native/quad_tree_lod_generic.gd [hunk 5]
```diff
@@ -115,7 +109,7 @@ func _update(quad: Quad, lod: int, view_pos: Vector3):
 			# Split
 			quad.children = [null, null, null, null]
 
-			for i in 4:
+			for i in range(4):
```
Comment (Zylann) [inline addons/zylann.hterrain/native/quad_tree_lod_generic.gd:112]: `range` is not necessary

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+	if (! _tree->has_children() && _tree->data.get_type() == Variant::NIL)
+		_tree->data = _make_chunk(_max_depth, 0, 0);
+}
+
+void QuadTreeLod::_update(Quad *quad, int lod, Vector3 view_pos) {
+    // This function should be called regularly over frames.
+
+    int lod_factor = get_lod_factor(lod);
+    int chunk_size = _base_size * lod_factor;
+    Vector3 world_center = (real_t)chunk_size * (Vector3((real_t)quad->origin_x, 0, (real_t)quad->origin_y) + Vector3(0.5, 0, 0.5));
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:103]: Perhaps it is not necessary to cast so explicitely in the C-style way, at least in the parameters to Vector3

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+        int checker = 0;
+        if (child_index == 1 || child_index == 2)
+            checker = 1;
+
+        int chunk_indicator = 0;
+        if (quad->data.get_type() == Variant::OBJECT)
+            chunk_indicator = 1;
+
+        Variant rect2(Rect2( Vector2((real_t)quad->origin_x, (real_t)quad->origin_y) * (real_t)size, 
+                             Vector2((real_t)size, (real_t)size)));
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:212]: Casting in C-style may not be necessary when sending arguments to Vector2 constructor

File: addons/zylann.hterrain/native/src/quad_tree_lod.h [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+#include <FuncRef.hpp>
+
+namespace godot {
+
+class QuadTreeLod : public Reference {
+    GODOT_CLASS(QuadTreeLod, Reference)
+
+    class Quad {
+    private:
+        Quad **children = nullptr;
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.h:14]: This is a dynamically-allocated array for fixed amount of items, which itself contains dynamically-allocated quads. In GDScript there was no other choice, but in C++ it is possible to rather use a single vector of Quads like a pool where every pack of 4 is used for children. An index to the first child can be used instead of a pointer, and the boolean can be removed by interpreting a negative index as having no children. This way we go from 16-byte structs scattered on the heap to 4-byte structs close together in memory (it has its own tricky parts, but I've done that with an octree: https://github.com/Zylann/godot_voxel/blob/master/terrain/lod_octree.h). It might bring minor improvement in 2D but it could help. I dont think this causes your slowdown though.

**Review by Zylann — COMMENTED** [review]
I did a bit of a closer read, could not figure out what could cause a slowdown.

**Comment by tinmanjuggernaut** [comment]
Thanks for taking a look at the code.

> Note, I completely stopped doing things with GDNative as I am waiting for GDExtensions in Godot 4 now, which may be a better system.

I understand. Though I'm working on a game now with GD3. I hope to be finished long before GD4 comes out.

> * Obvious question would be, did you test with optimizations enabled? (i.e no debug)

I thought I was, but apparently, it defaults to a debug target. I recompiled godot-cpp and this package with target=release and confirmed the compiler flags had /o2. However it doesn't make much difference.

> * Note that the slowness may not entirely come from QuadTreeLod alone, the logic found in callbacks may be responsible too.
The call backs refer to the gdscript code. So it should be slow only from my code, the godot-cpp interface, any code that transfers objects and data between GDScript and C++ via a Variant. The remaining gdscript hasn't changed so shouldn't be slower, right?

Here you can see gdnative above at half the speed and stuck on a higher LOD. Forcing the factory to return gdscript is the only difference between these two runs. Maybe the logic bug that is causing this is also the cause for the slowdown?
image

> * If I were to port this to C++, approaches from GDScript don't apply: for example, I would not use heap-allocated quads like you did, it's better if you use a packed pool with one vector, and indexing them by groups of 4 instead, much better memory locality and less pointers

I was hoping to see an improvement from a 1:1 conversion and refactor algorithms later. But since it doesn't work properly, I will try it now. Maybe that will unearth the LOD bug. I believe I understand what you mean about a single vector for all of the quads, each has an index into the vector to their child group. Should I make `_tree` an `Array` or a std::vector?

Thanks

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 6 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:103]: Every (real_t) cast I put in was just to get rid of compiler warnings. All of these variables are ints being passed to real_t functions.

File: addons/zylann.hterrain/native/quad_tree_lod_generic.gd (hunk 5 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/quad_tree_lod_generic.gd:112]: I did this only for consistency. The file had 3 methods of iteration: `4`, `range(4)`, and `quad.children.size()`. I could change them all to `4`.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 2 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:172]: `!= null`, aka `is valid`. Since data is a Variant, it's always valid.
I changed these to `quad->data.get_type() != Variant::NIL` but no difference.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 1 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:120]: Yes, I see that now. However attempting to get_child(i) and using it caused segfaults. In any case, I will convert the code to the vector method.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 6 above)
Comment (Zylann) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:103]: I wonder where these warnings come from. I never get these when I pass ints to functions expecting floats when working on the engine 🤔

**Comment by Zylann** [comment]
I think it is likely that the functional issue you are having is also causing the slowdown.

> Should I make _tree an Array or a std::vector

In my octree implementation I used two vectors. One to store quads, and another to store free indices, because after releasing a quad, I actually don't free it, I just leave it where it is in the vector, and instead push the first child index into the list of free indices. So that when a ne set of quads is needed, I first look into the list of free ones. If that list is empty, only then I push 4 new quads at the end of the vector.
One detail to be careful with this approach is that any code that potentially creates new quads can invalidate pointers/refs to quads, which means sometimes the ref must be re-obtained (like here: https://github.com/Zylann/godot_voxel/blob/5959699b9b5726f612cb76eeb0df8be44b32046d/terrain/lod_octree.h#L214)
That said, given the quad tree is 2D, you might be able to get away with preallocating the maximum amount of quads up-front. It would occupy more memory, but it eliminates the case.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 6 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:103]: Your SConstruct lists -W3 which includes `lint-like warnings`.
Godot's sconstruct defaults to no warnings, /W2 if `moderate` is requested, /W3 for `all` and /Wall for `extra`

**Comment by tinmanjuggernaut** [comment]
In the posted version, _update was never called with lower LODs. The way I determined if a node had children was wrong as that branch of _update never ran.

Making a version similar to your octree, using a prebuilt QTL vector, I was able to get it working. Turning off all detail layers and stripping down the scene to just the player, landscape, and environment my new version gets about a 5-20% improvement depending on where you're looking. e.g. 180->190fps to 170->200fps. 

That's enough of a speed benefit that it's probably worth hitting a few more classes, like the detail layers. Many other factors slow down my game, but at least I can make sure the terrain is out of the way. Next, I'll clean up what I have and post it. Then tackle a few other worst offenders.

I also noticed in both the gdscript version and C++ version that static memory gradually increases by 0.01MB every 1-few seconds as I move around, but not when I stand still.

**Comment by Zylann** [comment]
> I also noticed in both the gdscript version and C++ version that static memory gradually increases by 0.01MB every 1-few seconds as I move around, but not when I stand still.

It is possible some things are getting cached the first time you move into an area of the terrain. To be sure you would have to move through every location and see if it still increases afterward.

**Comment by tinmanjuggernaut** [comment]
I've cleaned up and pushed the working native QTL. 
Notes:

https://github.com/Zylann/godot_heightmap_plugin/blob/34a02792d03a46197ec27292722e81f0d3b16ff9/addons/zylann.hterrain/util/quad_tree_lod.gd#L140
Lines 140, 144, 147 are redundant so I have removed them in both gd/c++ versions. Quad was already checked for children, quad.data got reset 4 times in the loop, then again on 148.

Now both versions seem to run even faster. Peaking at 230fps on the C++ version that only got up to about 200-210fps yesterday.

Feel free to identify anything on the QTL that needs to be corrected. I'll start working on another class.

File: addons/zylann.hterrain/native/src/quad_tree_lod.h [hunk 9]
```diff
@@ -3,120 +3,125 @@
...
+            origin_y = 0;
+            clear_data();
         }
 
-        void add_children() {
-            if (_has_children)
-                Godot::print("Error: Quad::add_children(): already has children");
-                return;
+        inline void clear_data() {
+            _data = Variant();
```
Comment (tinmanjuggernaut) [inline addons/zylann.hterrain/native/src/quad_tree_lod.h:44]: Does this adequately remove our reference to the underlying object? I wasn't sure about it looking at the variant code. It seemed to me that I needed to run Variant::clear() to remove the reference, but that's not available in godot-cpp. The only way to get it is to call the destructor. But then I'd have to use pointers and new so I could delete _data, right?

Comment (Zylann) [reply addons/zylann.hterrain/native/src/quad_tree_lod.h:44]: I assume replacing with an empty variant is similar to assigning to `null` in GDScript.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+        quad->first_child = NO_CHILDREN;
+    }
+}
+
+// Returns the index of the first_child. Allocates from _free_indices.
+unsigned int QuadTreeLod::_allocate_children() {
+    if (_free_indices.size() == 0) {
+        unsigned int i0 = static_cast<unsigned int>(_tree.size());
+        _tree.resize(i0 + 4);
+        return 4 * i0;
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:76]: Why are you multiplying the index by 4 🤔

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+// Pass the first_child index, not the parent index. Stores back in _free_indices.
+void QuadTreeLod::_recycle_children(unsigned int i0) {
+    // Debug check, there is no use case in recycling a node which is not a first child
+    CRASH_COND(i0 % 4 != 0);
+
+    for (int i = 0; i < 4; ++i) {
+        _tree[i0 + i].init();
+    }
+
+    _free_indices.push_back(i0 / 4);
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:93]: And why divide by 4? Seems like you are working with a different type of index here, but it's really unnecessary. You can totally store actual indexes to first children in there, still without storing the 3 others

File: addons/zylann.hterrain/native/src/quad_tree_lod.h [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+        inline void set_data(Variant p_data) {
+            _data = p_data;
+        }
+
+    };
+
+private:
+    Quad _root;
+    std::vector<Quad> _tree;
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.h:71]: This is technically not a tree but a pool of quads, not sure if the name is accurate

**Review by Zylann — COMMENTED** [review]
Did you test with a C++ debugger in case there are unseen exceptions?

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+    }
+}
+
+void QuadTreeLod::create_from_sizes(int base_size, int full_size) {
+    clear();
+    _base_size = base_size;
+    _max_depth = compute_lod_count(base_size, full_size);
+
+    int node_count = static_cast<int>(pow(4, _max_depth) / (4 - 1)) - 1;    // # nodes is N^L / (N - 1). -1 for root, where N=num children, L=levels
+    _tree.resize(node_count);                               // e.g. (4^6 / 3 ) - 1 = 1364(.33) excluding root
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:138]: I think you dont actually want `resize`, but `reserve`. If you want to use `resize`, then you should never use it again afterward because it's already the biggest size possible, no need to shrink it or expand it. It seems you still copied some dynamic logic from the octree, which was not preallocating nodes

**Comment by tinmanjuggernaut** [comment]
Hey, I've been preoccupied but will cycle back on this soon.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 13 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:138]: Ah yes, that's exactly what I did. I duplicated your octree code, and also preallocating everything. I'll fix the function.

File: addons/zylann.hterrain/native/src/quad_tree_lod.h (hunk 12 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.h:71]: renamed

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 11 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:93]: We've adjusted this. _free_indices is still only 1/4th the size, and it stores only the multiples of 4: [ 0, 4, 8, 16].

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 10 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:76]: Yes, that was my mistake, and this code should never be reached anyway. We've removed it.

File: addons/zylann.hterrain/native/src/quad_tree_lod.h (hunk 9 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.h:44]: Ok, I won't worry about it then.

**Comment by tinmanjuggernaut** [comment]
Alright, I got back on this and am working with @painfulexistence to finish this up. He fixed the bugs you noted, then we reviewed it together. I think this PR is fine or just the QTL and other classes can be looked at later.

I think it's ready for your review. It has only a few minor updates since your last review.

**Comment by tinmanjuggernaut** [comment]
@Zylann This PR has been ready for your review for the last month. It's been working fine in our game and the code hasn't changed. I just pushed an update with a linux binary. Let me know if you'd like any changes to the PR.

FYI, @painfulexistence has been reworking the detail layers class design in gdscript and porting it to gdnative.

**Comment by Zylann** [comment]
Will have a last look when I get the time.
I realize in the repo we have a 16-month old submodule of godot-cpp, it's quite interesting we did not need to update it since.

File: addons/zylann.hterrain/native/factory.gd [hunk 14]
```diff
@@ -26,3 +27,12 @@ static func get_image_utils():
...
 			return ImageUtilsNative.new()
 	return ImageUtilsGeneric.new()
+
+
+static func get_quad_tree_lod():
+	if is_native_available():
+		var QuadTreeLod = load(NATIVE_PATH + "quad_tree_lod.gdns")
+		if QuadTreeLod != null:
+			return QuadTreeLod.new()
+	else:
```
Comment (Zylann) [inline addons/zylann.hterrain/native/factory.gd:37]: There should be no `else`

File: addons/zylann.hterrain/native/quad_tree_lod_generic.gd [hunk 15]
```diff
@@ -70,18 +62,49 @@ func set_split_scale(p_split_scale: float):
...
-
-	_split_scale = float(p_split_scale)
+	_split_scale = clamp(p_split_scale, MIN, MAX)
 
 
 func get_split_scale() -> float:
 	return _split_scale
 
 
+func clear():
```
Comment (Zylann) [inline addons/zylann.hterrain/native/quad_tree_lod_generic.gd:72]: Why did you move these down?

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    int chunk_size = _base_size * lod_factor;
+    Vector3 world_center = static_cast<real_t>(chunk_size) * (Vector3(static_cast<real_t>(quad->origin_x), 0.f, static_cast<real_t>(quad->origin_y)) + Vector3(0.5f, 0.f, 0.5f));
+
+    if (_vertical_bounds_func.is_valid()) {
+        Variant result = _vertical_bounds_func->call_func(quad->origin_x, quad->origin_y, lod);
+        Vector2 vbounds;
+        if (result.get_type() == Variant::VECTOR2)
+            vbounds = static_cast<Vector2>(result);
+        else
+            Godot::print("Error: QuadTreeLod::_update: _vertical_bounds_func did not return a Vector2.");
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:167]: Could use `ERR_FAIL_COND`

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 17]
```diff
@@ -0,0 +1,248 @@
...
+        world_center.y = (vbounds.x + vbounds.y) / 2.0f;
+    }
+
+    int split_distance = _base_size * lod_factor * static_cast<int>(_split_scale);
+
+    if (! quad->has_children()) {
+        if (lod > 0 && world_center.distance_to(view_pos) < split_distance) {
+            // Split
+            unsigned int new_idx = _allocate_children();
+            if (new_idx == NO_CHILDREN) {
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:177]: Could use `ERR_FAIL_COND`

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 18]
```diff
@@ -0,0 +1,248 @@
...
+        quad->clear_data();
+    }
+}
+
+void QuadTreeLod::create_from_sizes(int base_size, int full_size) {
+    clear();
+    _base_size = base_size;
+    _max_depth = compute_lod_count(base_size, full_size);
+
+    int node_count = static_cast<int>(pow(4, _max_depth) / (4 - 1)) - 1;    // # nodes is N^L / (N - 1). -1 for root, where N=num children, L=levels
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:135]: We should be lucky `pow` uses `double`, because `float` might be unable to accurately contain some of the biggest results this can get. Or we could try using an integer version

**Review by Zylann — COMMENTED** [review]
Maybe at this point we could include windows builds to the CI so they dont have to be made manually

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 18 above)
Comment (Zylann) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:135]: Thinking back, the maximum size in the plugin at the moment is 4096, and minimum chunk size is 16. Considering `Quad` is `(32+4+4+4)` bytes, that means the maximum amount of nodes we can have is:
```
>>> (32+4+4+4) * (int(pow(4, 9) / (4 - 1)) - 1)
3844720
```
3.8 megabytes is quite a lot for a preallocation strategy. It also will never get reached, unless you bump LOD distance such that the entire terrain is visible at LOD0.
Higher chunk size (so less quads) is still recommended, but maybe in the future we'll switch away from preallocating, depending on design changes.

File: addons/zylann.hterrain/native/factory.gd (hunk 14 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/factory.gd:37]: done

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 16 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:167]: done

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 18 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp:135]: I don't see any change requests here, so I won't make any for now.

File: addons/zylann.hterrain/native/quad_tree_lod_generic.gd (hunk 15 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/quad_tree_lod_generic.gd:72]: I matched the CPP version and organized the functions that call each other together. Clear, join all recursively, create_sizes, was one group. Update, _update, another. All of the smaller, individual functions at the top. Debug funcs at the bottom. 

Do you want me to revert to the original order? Or just move a few functions?

After you specify that, I'll rebuild and push all the changes.

Comment (Zylann) [reply addons/zylann.hterrain/native/quad_tree_lod_generic.gd:72]: Having them match is not enforced. But the main issue is that you mixed "formatting" with logic changes, it makes it harder to review code. Preferably, formatting should be in separate commits.

Also, the C++ class is not meeting the standards I use (see ImageUtils, and [this](https://voxel-tools.readthedocs.io/en/latest/module_development/#code-guidelines)), however as I said earlier I'm no longer working with GDNative as there will probably be a big rewrite (not just a port) of the plugin with GDExtension in the future, so it's not a big deal

**Comment by tinmanjuggernaut** [comment]
This new version:
* removes gdscript `else`
* Formats QTL w/ clang-format.
* Reorganizes public/private functions in QTL
* Uses ERR_FAIL_COND instead of Godot::print
* Reverts formatting of gdscript QTL
* windows/linux binaries rebuilt and tested

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 19]
```diff
@@ -0,0 +1,248 @@
...
+	int chunk_size = _base_size * lod_factor;
+	Vector3 world_center = static_cast<real_t>(chunk_size) * (Vector3(static_cast<real_t>(quad->origin_x), 0.f, static_cast<real_t>(quad->origin_y)) + Vector3(0.5f, 0.f, 0.5f));
+
+	if (_vertical_bounds_func.is_valid()) {
+		Variant result = _vertical_bounds_func->call_func(quad->origin_x, quad->origin_y, lod);
+		Vector2 vbounds;
+		if (result.get_type() == Variant::VECTOR2)
+			vbounds = static_cast<Vector2>(result);
+		else
+			ERR_FAIL_COND("_vertical_bounds_func did not return a Vector2.");
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:159]: Invalid use of `ERR_FAIL_COND`: the way it works is to take the failing condition as argument.
So the code would have to be:
```cpp
Variant result = _vertical_bounds_func->call_func(quad->origin_x, quad->origin_y, lod);
ERR_FAIL_COND(result.get_type() != Variant::VECTOR2);
Vector2 vbounds = static_cast<Vector2>(result);
```
There is also a version taking an extra message if you wish:
```cpp
ERR_FAIL_COND_MSG(result.get_type() != Variant::VECTOR2, "Vertical bounds function must return a Vector2");
```
And when the passed condition is true, it internally does `return;`, aborting the function.
The compiler appears to have accepted it because technically a string is booleanizable to `true`, but that's a misuse of the macro nevertheless.

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp [hunk 20]
```diff
@@ -0,0 +1,248 @@
...
+	}
+
+	int split_distance = _base_size * lod_factor * static_cast<int>(_split_scale);
+
+	if (!quad->has_children()) {
+		if (lod > 0 && world_center.distance_to(view_pos) < split_distance) {
+			// Split
+			unsigned int new_idx = _allocate_children();
+			if (new_idx == NO_CHILDREN) {
+				ERR_FAIL_COND("_free_indices is empty.");
```
Comment (Zylann) [inline addons/zylann.hterrain/native/src/quad_tree_lod.cpp:170]: Other misuse of the macro:
```cpp
ERR_FAIL_COND_MSG(new_idx == NO_CHILDREN, "_free_indices is empty.");
```

File: addons/zylann.hterrain/native/src/quad_tree_lod.cpp (hunk 19 above)
Comment (tinmanjuggernaut) [reply addons/zylann.hterrain/native/src/quad_tree_lod.cpp]: Ah of course, my mistake. This has been fixed.

**Comment by tinmanjuggernaut** [comment]
New build pushed.

**Comment by tinmanjuggernaut** [comment]
Wait, I noticed I get can't allocate nodes on 256 terrains, but it's fine on 512+. I think the number of nodes in the tree calculation is wrong.

**Comment by tinmanjuggernaut** [comment]
Alright, this is ready to go again.

I fixed the allocation calculation here: 
https://github.com/Zylann/godot_heightmap_plugin/blob/3d31072380ac2cc48a7c8263d2877cd2e39e7d20/addons/zylann.hterrain/native/src/quad_tree_lod.cpp#L61

The number of nodes in the tree is N^L/N-1, where L is the number of levels including the root level. `compute_lod_count` returns the number of levels minus the root level, so we must add one more to max_depth, then later subtract the one node for the root.

With a chunk size of 32, a 64x64 terrain has 4 nodes. I stepped through and tested all of the resolution sizes and got these node counts:
* 65: 4
* 127: 20
* 257: 84
* 513: 340
* 1025: 1364
* 2049: 5460
* 4097: 21,844
* 4097 @ 16: 87,380 * 44 bytes = 3.8mb as you calculated.

I also ran a player/environment/terrain with the last two and both ran at 200-300+fps on my GTX 1060. I didn't notice any performance difference between them. However I notice a significant performance improvement under Linux compared to Windows.
